text/0000-safer-transmute.md

```diff
@@ -0,0 +1,1796 @@
+
+
+### 📖 When is a transmutation sound?
+[sound transmutation]: #-when-is-a-transmutation-sound
+A transmutation is ***sound*** if the mere act of transmuting a value from one type to another is not unspecified or undefined behavior.
+
+#### Well-Defined Representation
+[`u8`]: core::u8
+[`f32`]: core::f32
+
+Transmutation is *always unsound* if it occurs between types with unspecified representations.
+
+Most of Rust's primitive types have specified representations. That is, the precise layout characteristics of [`u8`], [`f32`] is a documented and guaranteed aspect of those types.
+
+In contrast, most `struct` and `enum` types defined without an explicit `#[repr(C)]` attribute do ***not*** have well-specified layout characteristics.
+
+To ensure that types you've define are soundly transmutable, you almost always (with very few exceptions) must mark them with the `#[repr(C)]` attribute.
+
+#### Requirements on Owned Values
+[transmute-owned]: #requirements-on-owned-values
+
+Transmutations involving owned values must adhere to two rules to be sound. They must:
+ * [preserve or broaden the bit validity][owned-validity], and
+ * [preserve or shrink the size][owned-size].
+
+##### Preserve or Broaden Bit Validity
+[owned-validity]: #Preserve-or-Broaden-Bit-Validity
+[`NonZeroU8`]: https://doc.rust-lang.org/beta/core/num/struct.NonZeroU8.html
+
+The bits of any valid instance of the source type must be a bit-valid instance of the destination type.
+
+For example, we are permitted to transmute a `Bool` into a [`u8`]:
+```rust
+#[derive(Default, PromiseTransmutableFrom, PromiseTransmutableInto)]
+#[repr(u8)]
+enum Bool {
+    True = 1,
+    False = 0,
+}
+
+let _ : u8 = Bool::True.transmute_into();
+let _ : u8 = Bool::False.transmute_into();
+```
+<sup>
+(Note: <code>#[derive(PromiseTransmutableFrom, PromiseTransmutableInto)]</code> annotation connotes that <i>all</i> aspects of <code>Bool</code>'s layout are part of its <a href="#-when-is-a-transmutation-stable">library stability guarantee</a>.)
+</sup>
+
+...because all possible instances of `Bool` are also valid instances of [`u8`]. However, transmuting a [`u8`] into a `Bool` is forbidden:
+```rust
+/* ⚠️ This example intentionally does not compile. */
+let _ : Bool = u8::default().transmute_into(); // Compile Error!
+```
+...because not all instances of [`u8`] are valid instances of `Bool`.
+
+Another example: While laying out certain types, Rust may insert padding bytes between the layouts of fields. In the below example `Padded` has two padding bytes, while `Packed` has none:
```

Comment:
all values of `* T` are valid values of `* U` assuming that `T` and `U` are `Sized`, since a pointer is not required to be aligned.

Reply:
> Transmuting from a *T to a &T, because *T can be null (0) but &T can't.

And because `&T` must be aligned and dereferencable and point to a valid `T`. There are many conditions on references.^^